snappy/security.go

```diff
@@ -20,94 +20,241 @@
+// defaults
+type SecurityAppArmorOverrideDefinition struct {
+	ReadPaths    []string `yaml:"read-paths,omitempty" json:"read-paths,omitempty"`
+	WritePaths   []string `yaml:"write-paths,omitempty" json:"write-paths,omitempty"`
+	Abstractions []string `yaml:"abstractions,omitempty" json:"abstractions,omitempty"`
+}
+
+// SecurityOverrideDefinition is used to override apparmor or seccomp
+// security defaults
+type SecurityOverrideDefinition struct {
+	AppArmor *SecurityAppArmorOverrideDefinition `yaml:"apparmor" json:"apparmor"`
+	Seccomp  *SecuritySeccompOverrideDefinition  `yaml:"seccomp" json:"seccomp"`
+}
+
+// SecurityPolicyDefinition is used to provide hand-crafted policy
+type SecurityPolicyDefinition struct {
+	AppArmor string `yaml:"apparmor" json:"apparmor"`
+	Seccomp  string `yaml:"seccomp" json:"seccomp"`
+}
+
+// SecurityDefinitions contains the common apparmor/seccomp definitions
+type SecurityDefinitions struct {
+	// SecurityTemplate is a template like "default"
+	SecurityTemplate string `yaml:"security-template,omitempty" json:"security-template,omitempty"`
+	// SecurityOverride is a override for the high level security json
+	SecurityOverride *SecurityOverrideDefinition `yaml:"security-override,omitempty" json:"security-override,omitempty"`
+	// SecurityPolicy is a hand-crafted low-level policy
+	SecurityPolicy *SecurityPolicyDefinition `yaml:"security-policy,omitempty" json:"security-policy,omitempty"`
+
+	// SecurityCaps is are the apparmor/seccomp capabilities for an app
+	SecurityCaps []string `yaml:"caps,omitempty" json:"caps,omitempty"`
+}
+
+// securityPolicyType is a kind of securityPolicy, we currently
+// have "apparmor" and "seccomp"
+type securityPolicyType struct {
+	name          string
+	basePolicyDir string
+}
+
+var securityPolicyTypeAppArmor = securityPolicyType{
+	name:          "apparmor",
+	basePolicyDir: "/usr/share/apparmor/easyprof",
+}
+
+var securityPolicyTypeSeccomp = securityPolicyType{
+	name:          "seccomp",
+	basePolicyDir: "/usr/share/seccomp",
+}
+
+func (sp *securityPolicyType) policyDir() string {
+	return filepath.Join(dirs.GlobalRootDir, sp.basePolicyDir)
+}
+
+func (sp *securityPolicyType) frameworkPolicyDir() string {
+	frameworkPolicyDir := filepath.Join(policy.SecBase, sp.name)
+	return filepath.Join(dirs.GlobalRootDir, frameworkPolicyDir)
+}
 
-	// FIXME: this is snappy specific, on other systems like the
-	//        phone we may want different defaults.
-	if t.Template == "" && t.PolicyGroups == nil {
-		t.PolicyGroups = defaultPolicyGroups
+func (sp *securityPolicyType) findTemplate(template string) (string, error) {
+	if template == "" {
+		template = defaultTemplate
 	}
 
-	// never write a null value out into the json
-	if t.PolicyGroups == nil {
-		t.PolicyGroups = []string{}
+	systemTemplate := ""
+	fwTemplate := ""
+	subdir := filepath.Join("templates", defaultPolicyVendor(), defaultPolicyVersion())
+	systemTemplate = filepath.Join(sp.policyDir(), subdir, template)
```

Comment:
It would be nice to clean up nomenclature a little bit if possible. Reading through the code, template names, template paths, and template contents, are all "template".